### PR TITLE
Add `metadata-file` flag

### PR DESCRIPTION
**Pull request by crazy-max**
cc. @tonistiigi 

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

### CODE REVIEW

**Comment by tonistiigi** [comment]
I think this would be better as metadata-file (eg. https://github.com/moby/buildkit/pull/2095) and one per bake that lists all properties for all built targets instead of a file per target.

@AkihiroSuda thoughts?

**Comment by AkihiroSuda** [comment]
SGTM

**Review by tonistiigi — CHANGES_REQUESTED** [review]
^

**Comment by crazy-max** [comment]
Ok here is the current state on my side:

* `build` command
  * `--iidfile` same behavior
  * `--metadatafile` write `ExporterResponse` (single target)
* `bake` command
  * `--iidfile` remove this flag introduced here in favor of the new flag `--metadatafile`
  * `--metadatafile` write all build options and metadata for all targets. In this case I don't think we should handle a `metadatafile` target field.

**Comment by tonistiigi** [comment]
@crazy-max SGTM. Having both `iidfile` and `metadata-file` in build may be a bit confusing atm but I guess it is fine. We could deprecate `iidfile` later or make it hidden.

https://github.com/moby/buildkit/pull/2095/files added flag with a dash. I think the dash is better.

**Comment by crazy-max** [comment]
@tonistiigi 

> Having both iidfile and metadata-file in build may be a bit confusing atm but I guess it is fine. We could deprecate iidfile later or make it hidden.

Agree about deprecation of `iidfile`.

> https://github.com/moby/buildkit/pull/2095/files added flag with a dash. I think the dash is better.

Yes, I was wondering if we were keeping the same format as for iidfile but since we could depreciate it I will modify it to `metadata-file`.

**Comment by crazy-max** [comment]
@tonistiigi Ok now `metadata-file` handles build/solve. Here is an example:

```hcl
variable "DEFAULT_TAG" {
  default = "7zip:local"
}

// Special target: https://github.com/docker/metadata-action#bake-definition
target "docker-metadata-action" {
  tags = ["${DEFAULT_TAG}"]
}

// Default target if none specified
group "default" {
  targets = ["image-local"]
}

target "image" {
  inherits = ["docker-metadata-action"]
}

target "image-local" {
  inherits = ["image"]
  output = ["type=docker"]
}

target "image-all" {
  inherits = ["image"]
  platforms = [
    "linux/amd64",
    "linux/arm/v6",
    "linux/arm/v7",
    "linux/arm64",
    "linux/386",
    "linux/ppc64le",
    "linux/s390x"
  ]
}
```

```shell
docker buildx bake --metadata-file metadata.json image-local
```

> `metadata.json`
```json
{
  "metadata": {
    "build-opts": {
      "inputs": {
        "context": ".",
        "dockerfile": "Dockerfile"
      },
      "tags": [
        "7zip:local"
      ],
      "metadata-file": "./metadata.json"
    },
    "solve-opts": [
      {
        "containerimage.config.digest": "sha256:fe4387b1cf24e8b3f153b026e2843e8abddeda479b27fd6adf3ed7ad1f8eb67a",
        "containerimage.digest": "sha256:3eaf8fb892e653224be132bc26de3bfa82e4563750b5f1f6dd64ae25788c6247",
        "image.name": "docker.io/library/7zip:local"
      }
    ]
  }
}
```

Let me know if it's the right direction.

**Comment by tonistiigi** [comment]
I thought this was just going to contain the build result metadata. Why is that part called `solve-opts`? The rest is somewhat interesting but a different use-case and format is an internal representation that doesn't match anything else.

For the bake, the key should be `image-local`, not "metadata?"

Not related to this PR code but what is `containerimage.config.digest` for a multi-platform build.

**Comment by crazy-max** [comment]
@tonistiigi 

> I thought this was just going to contain the build result metadata. Why is that part called `solve-opts`?

Yes I was wondering about naming. Will call it `metadata` and set a "per-target" representation.

> The rest is somewhat interesting but a different use-case and format is an internal representation that doesn't match anything else.

I wanted to see if it could be interesting in the future indeed. We can do this in a follow-up instead.

> For the bake, the key should be `image-local`, not "metadata?"

Ok I can make a representation per-target

> Not related to this PR code but what is `containerimage.config.digest` for a multi-platform build.

Will be empty:

```shell
docker buildx bake --progress plain --set *.tags=crazymax/7zip:test --push --metadata-file ./metadata.json image-all
```

```json
{
  "metadata": {
    "build-opts": {
      "inputs": {
        "context": ".",
        "dockerfile": "Dockerfile"
      },
      "tags": [
        "crazymax/7zip:test"
      ],
      "metadata-file": "./metadata.json",
      "platform": [
        {
          "architecture": "amd64",
          "os": "linux"
        },
        {
          "architecture": "arm",
          "os": "linux",
          "variant": "v6"
        },
        {
          "architecture": "arm",
          "os": "linux",
          "variant": "v7"
        },
        {
          "architecture": "arm64",
          "os": "linux"
        },
        {
          "architecture": "386",
          "os": "linux"
        },
        {
          "architecture": "ppc64le",
          "os": "linux"
        },
        {
          "architecture": "s390x",
          "os": "linux"
        }
      ]
    },
    "solve-opts": [
      {
        "containerimage.digest": "sha256:61ba426e49e207972d9ffaf6761cffec9fb3ae63dab76414cf250c9058370643",
        "image.name": "crazymax/7zip:test"
      }
    ]
  }
}
```

**Comment by crazy-max** [comment]
Here are the new changes:

## `build`

```shell
docker buildx build --metadata-file metadata.json .
```

```json
{
  "metadata": [
    {
      "containerimage.config.digest": "sha256:fe4387b1cf24e8b3f153b026e2843e8abddeda479b27fd6adf3ed7ad1f8eb67a",
      "containerimage.digest": "sha256:3eaf8fb892e653224be132bc26de3bfa82e4563750b5f1f6dd64ae25788c6247"
    }
  ]
}
```

## `bake`

```shell
docker buildx bake --progress plain --set *.tags=crazymax/7zip:test --push --metadata-file metadata.json image-all
```

> `metadata.json`
```json
{
  "image-all": {
    "definition": {
      "context": ".",
      "dockerfile": "Dockerfile",
      "tags": [
        "crazymax/7zip:test"
      ],
      "platforms": [
        "linux/amd64",
        "linux/arm/v6",
        "linux/arm/v7",
        "linux/arm64",
        "linux/386",
        "linux/ppc64le",
        "linux/s390x"
      ],
      "output": [
        "type=registry"
      ]
    },
    "metadata": [
      {
        "containerimage.digest": "sha256:61ba426e49e207972d9ffaf6761cffec9fb3ae63dab76414cf250c9058370643",
        "image.name": "crazymax/7zip:test"
      }
    ]
  }
}
```

Don't handle multi bake targets in metadata result atm:

```hcl
variable "DEFAULT_TAG" {
  default = "7zip:local"
}

// Special target: https://github.com/docker/metadata-action#bake-definition
target "docker-metadata-action" {
  tags = ["${DEFAULT_TAG}"]
}

// Default target if none specified
group "default" {
  targets = ["image-local"]
}

target "image" {
  inherits = ["docker-metadata-action"]
}

target "image-local" {
  inherits = ["image"]
  output = ["type=docker"]
}

target "image-all" {
  inherits = ["image"]
  tags = ["crazymax/7zip:test"]
  output = ["type=registry"]
  platforms = [
    "linux/amd64",
    "linux/arm/v6",
    "linux/arm/v7",
    "linux/arm64",
    "linux/386",
    "linux/ppc64le",
    "linux/s390x"
  ]
}
```

```shell
docker buildx bake --metadata-file ./metadata.json image-all image-local --print
```

```json
{
   "target": {
      "image-all": {
         "context": ".",
         "dockerfile": "Dockerfile",
         "tags": [
            "crazymax/7zip:test"
         ],
         "platforms": [
            "linux/amd64",
            "linux/arm/v6",
            "linux/arm/v7",
            "linux/arm64",
            "linux/386",
            "linux/ppc64le",
            "linux/s390x"
         ],
         "output": [
            "type=registry"
         ]
      },
      "image-local": {
         "context": ".",
         "dockerfile": "Dockerfile",
         "output": [
            "type=docker"
         ]
      }
   }
}
```

```shell
docker buildx bake --metadata-file metadata.json image-all image-local
```

> `metadata.json`
```json
{
  "image-all": {
    "definition": {
      "context": ".",
      "dockerfile": "Dockerfile",
      "tags": [
        "crazymax/7zip:test"
      ],
      "platforms": [
        "linux/amd64",
        "linux/arm/v6",
        "linux/arm/v7",
        "linux/arm64",
        "linux/386",
        "linux/ppc64le",
        "linux/s390x"
      ],
      "output": [
        "type=registry"
      ]
    },
    "metadata": [
      {
        "containerimage.digest": "sha256:61ba426e49e207972d9ffaf6761cffec9fb3ae63dab76414cf250c9058370643",
        "image.name": "crazymax/7zip:test"
      }
    ]
  }
}
```

**Comment by crazy-max** [comment]
@tonistiigi As discussed I've removed the build definition and only outputs metadata (aka `ExporterReponse`).

```shell
docker buildx build --output type=docker --metadata-file ./metadata.json .
```

```json
{
  "metadata": {
    "containerimage.config.digest": "sha256:fe4387b1cf24e8b3f153b026e2843e8abddeda479b27fd6adf3ed7ad1f8eb67a",
    "containerimage.digest": "sha256:3eaf8fb892e653224be132bc26de3bfa82e4563750b5f1f6dd64ae25788c6247"
  }
}
```

____

```shell
docker buildx bake --metadata-file ./metadata.json image-all
```

```json
{
  "metadata": {
    "containerimage.digest": "sha256:61ba426e49e207972d9ffaf6761cffec9fb3ae63dab76414cf250c9058370643",
    "image.name": "crazymax/7zip:test"
  }
}
```

**Comment by crazy-max** [comment]
@tonistiigi Ok now targets are merged. Here is what it looks like

## `build`

```shell
docker buildx build --output type=docker --metadata-file ./metadata.json .
```

```json
{
  "metadata": {
    "default": {
      "containerimage.config.digest": "sha256:fe4387b1cf24e8b3f153b026e2843e8abddeda479b27fd6adf3ed7ad1f8eb67a",
      "containerimage.digest": "sha256:3eaf8fb892e653224be132bc26de3bfa82e4563750b5f1f6dd64ae25788c6247"
    }
  }
}
```

## `bake`

```shell
docker buildx bake --metadata-file metadata.json image-all image-local
```

```json
{
  "metadata": {
    "image-all": {
      "containerimage.digest": "sha256:61ba426e49e207972d9ffaf6761cffec9fb3ae63dab76414cf250c9058370643",
      "image.name": "crazymax/7zip:test"
    },
    "image-local": {
      "containerimage.config.digest": "sha256:fe4387b1cf24e8b3f153b026e2843e8abddeda479b27fd6adf3ed7ad1f8eb67a",
      "containerimage.digest": "sha256:3eaf8fb892e653224be132bc26de3bfa82e4563750b5f1f6dd64ae25788c6247",
      "image.name": "docker.io/library/7zip:local"
    }
  }
}
```

File: docs/reference/buildx_bake.md [hunk 1]
```diff
@@ -18,6 +18,7 @@ Build from a file
 | `--builder string` | Override the configured builder instance |
 | [`-f`](#file), [`--file stringArray`](#file) | Build definition file |
 | `--load` | Shorthand for --set=*.output=type=docker |
+| `--metadata-file string` | Write build metadata to a file as JSON |
```
Comment (tonistiigi) [inline docs/reference/buildx_bake.md:21]: Write build result metadata

**Comment by crazy-max** [comment]
@tonistiigi

## `build`

```shell
docker buildx build --output type=docker --metadata-file ./metadata.json .
```

```json
{
  "containerimage.config.digest": "sha256:d8b8b4f781520aeafedb5a88ff50fbb625cfebad87e392794f1e26a724a2f22a",
  "containerimage.digest": "sha256:868f04872380274dcf8528222e84dc66702394de80889e51c87a14126ea9ff6a"
}
```

## `bake`

```shell
docker buildx bake --metadata-file metadata.json image-all image-local
```

```json
{
  "image-all": {
    "containerimage.digest": "sha256:d9aeb67cb21c2b79da504342e3d30a6eb914b9d1b7df0c16b1947ee76587702c",
    "image.name": "crazymax/7zip:test"
  },
  "image-local": {
    "containerimage.config.digest": "sha256:d8b8b4f781520aeafedb5a88ff50fbb625cfebad87e392794f1e26a724a2f22a",
    "containerimage.digest": "sha256:868f04872380274dcf8528222e84dc66702394de80889e51c87a14126ea9ff6a",
    "image.name": "docker.io/library/7zip:local"
  }
}
```

File: build/build.go [hunk 2]
```diff
@@ -336,7 +336,7 @@ func toSolveOpt(ctx context.Context, d driver.Driver, multiDriver bool, opt Opti
 		}
 	}()
 
-	if opt.ImageIDFile != "" {
+	if len(opt.ImageIDFile) > 0 {
```
Comment (tonistiigi) [inline build/build.go:339]: hmm.., why this change(s)? The previous one is much clearer imho, and type safe.

File: commands/bake.go [hunk 3]
```diff
@@ -108,7 +109,28 @@ func runBake(dockerCli command.Cli, targets []string, in bakeOptions) (err error
...
 	}
 
-	_, err = build.Build(ctx, dis, bo, dockerAPI(dockerCli), dockerCli.ConfigFile(), printer)
+	resp, err := build.Build(ctx, dis, bo, dockerAPI(dockerCli), dockerCli.ConfigFile(), printer)
+	if err != nil {
+		return err
+	}
+
+	if len(in.metadataFile) > 0 && resp != nil {
+		if err := os.Remove(in.metadataFile); err != nil && !os.IsNotExist(err) {
```
Comment (tonistiigi) [inline commands/bake.go:118]: Could we do this with `os.Rename` instead or with `AtomicWriteFile()`. Is this for symlink protection? Otherwise, I don't think you need it at all.

File: docs/reference/buildx_build.md [hunk 4]
```diff
@@ -25,6 +25,7 @@ Start a build
 | `--iidfile string` | Write the image ID to the file |
 | `--label stringArray` | Set metadata for an image |
 | [`--load`](#load) | Shorthand for --output=type=docker |
+| `--metadata-file string` | Write build result metadata |
```
Comment (tonistiigi) [inline docs/reference/buildx_build.md:28]: ".. to the file" (matches iidfile)

File: commands/build.go [hunk 5]
```diff
@@ -213,11 +216,27 @@ func buildTargets(ctx context.Context, dockerCli command.Cli, opts map[string]bu
...
 	}
+	if err != nil {
+		return err
+	}
+
+	if len(metadataFile) > 0 && resp != nil {
+		if err := os.Remove(metadataFile); err != nil && !os.IsNotExist(err) {
+			return errors.Wrap(err, "removing metadata file")
+		}
+		mdatab, err := json.MarshalIndent(resp["default"].ExporterResponse, "", "  ")
```
Comment (tonistiigi) [inline commands/build.go:232]: define this "default" as constant now as used in multiple places.

**Review by tonistiigi — COMMENTED** [review]
Small comments. Overall looks good 👍

File: commands/bake.go (hunk 3 above)
Comment (crazy-max) [reply commands/bake.go]: > with `AtomicWriteFile()`

SGTM

File: build/build.go (hunk 2 above)
Comment (crazy-max) [reply build/build.go]: Reverted. Was using the same check for the previous impl for metadata-file, forgot to cleanup. It's more a matter of personal taste but Go compiler generates [identical code](https://godbolt.org/z/fib1x1).

File: build/build.go [hunk 6]
```diff
@@ -45,6 +45,10 @@ var (
 	errDockerfileConflict = errors.New("ambiguous Dockerfile source: both stdin and flag correspond to Dockerfiles")
 )
 
+const (
+	DefaultTargetName = "default"
```
Comment (tonistiigi) [inline build/build.go:49]: it can remain in commands and private if not used cross pkg

**Review by tonistiigi — APPROVED** [review]
Now that we import `ioutils` you can remove the duplicate function in `store` pkg in follow-up.
